tests/src/drivers/keygen.c

```diff
@@ -57,58 +57,8 @@ psa_status_t test_transparent_generate_key(
         return( PSA_SUCCESS );
     }
 
-    /* Copied from psa_crypto.c */
-#if defined(MBEDTLS_ECP_C)
-    if ( PSA_KEY_TYPE_IS_ECC( psa_get_key_type( attributes ) )
-         && PSA_KEY_TYPE_IS_KEY_PAIR( psa_get_key_type( attributes ) ) )
-    {
-        psa_ecc_family_t curve = PSA_KEY_TYPE_ECC_GET_FAMILY( psa_get_key_type( attributes ) );
-        mbedtls_ecp_group_id grp_id =
-            mbedtls_ecc_group_of_psa( curve, PSA_BITS_TO_BYTES( psa_get_key_bits( attributes ) ) );
-        const mbedtls_ecp_curve_info *curve_info =
-            mbedtls_ecp_curve_info_from_grp_id( grp_id );
-        mbedtls_ecp_keypair ecp;
-        mbedtls_test_rnd_pseudo_info rnd_info;
-        memset( &rnd_info, 0x5A, sizeof( mbedtls_test_rnd_pseudo_info ) );
-
-        int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
-        if( attributes->domain_parameters_size != 0 )
-            return( PSA_ERROR_NOT_SUPPORTED );
-        if( grp_id == MBEDTLS_ECP_DP_NONE || curve_info == NULL )
-            return( PSA_ERROR_NOT_SUPPORTED );
-        if( curve_info->bit_size != psa_get_key_bits( attributes ) )
-            return( PSA_ERROR_INVALID_ARGUMENT );
-        mbedtls_ecp_keypair_init( &ecp );
-        ret = mbedtls_ecp_gen_key( grp_id, &ecp,
-                                   &mbedtls_test_rnd_pseudo_rand,
-                                   &rnd_info );
-        if( ret != 0 )
-        {
-            mbedtls_ecp_keypair_free( &ecp );
-            return( mbedtls_to_psa_error( ret ) );
-        }
-
-        /* Make sure to use export representation */
-        size_t bytes = PSA_BITS_TO_BYTES( psa_get_key_bits( attributes ) );
-        if( key_size < bytes )
-        {
-            mbedtls_ecp_keypair_free( &ecp );
-            return( PSA_ERROR_BUFFER_TOO_SMALL );
-        }
-        psa_status_t status = mbedtls_to_psa_error(
-            mbedtls_mpi_write_binary( &ecp.d, key, bytes ) );
-
-        if( status == PSA_SUCCESS )
-        {
-            *key_length = bytes;
-        }
-
-        mbedtls_ecp_keypair_free( &ecp );
-        return( status );
-    }
-    else
-#endif /* MBEDTLS_ECP_C */
-    return( PSA_ERROR_NOT_SUPPORTED );
+    return( psa_generate_key_internal( attributes,
```

Comment:
This introduces an extra dependency of the test code on the library code. Before, the test code only depended on the mbedtls API plus the mbedtls/psa glue function `mbedtls_to_psa_error`. Now the code depends on code in `psa_crypto.c`.

If Mbed TLS is compiled without software support for e.g. ECC through the PSA API, then under the [new PSA configuration mechanism](https://github.com/ARMmbed/mbedtls/blob/development/docs/proposed/psa-conditional-inclusion-c.md), the PSA code in the Mbed TLS library should not include support for ECC, which means that `psa_generate_key_internal` would not support ECC, and therefore this driver would not be usable for testing. But a configuration where Mbed TLS doesn't have the software implementation is precisely where it's the most important to use the test driver. So we would need to add yet another testing mode so that the library code is included in the build. But then we wouldn't really be testing the real product.

Reply:
Replying to this comment on the first version of the PR for the sake of completeness. The main focus of the first version of the PR was the test driver code and to avoid to rewrite in test drivers psa_crypto,c. The scope of the changes have changed since and their rationale explained in docs/architecture/psa-crypto-implementation-structure.md. Even if I think that this PR already improve significantly the testing of the `generate_key` entry point and the dispatch of key generation to a transparent driver (see the PR description), it still does not address the root issue (tracked by issue #3680) raised by this comment that the test driver `generate_key` entry points code is part of the MbedTLS crypto library.

Reply:
This comment dates from before we settled on a test strategy that will involve throwing away much of the code in this file, right? Since this is short-term code, the extra dependency doesn't matter after all.